CLBlast support

**Is your feature request related to a problem? Please describe.**
No support for OpenCL.

**Describe the solution you'd like**
Implementation of CLBlast to provide support for OpenCL.

**Describe alternatives you've considered**
None, as implementing something for ROCm (if it was something more dedicated) would leave away pre-RDNA graphics cards for AMD, several integrated GPU for Intel, etc.

**Additional context**
Mostly wanted for AMD and Intel iGPUs.

I think it would be good if the instructions around getting CLBLAST to work were updated.

So far on Fedora I can't get LocalAI with CLBlast to build as the make fails with ld complaining about missing libraries - despite the build taking place within the official docker container and with CLBLAST installed, mounted as a volume and with CLBLAST_DIR set.

```
...
CGO_LDFLAGS="-lOpenCL -lclblast" C_INCLUDE_PATH=/build/go-ggml-transformers LIBRARY_PATH=/build/go-ggml-transformers \
go build -ldflags "-X "github.com/go-skynet/LocalAI/internal.Version=v1.22.0-10-g0883d32" -X "github.com/go-skynet/LocalAI/internal.Commit=0883d324d9b29b12e8417aa20d6458a77f62aab1"" -tags "" -o backend-assets/grpc/falcon-ggml ./cmd/grpc/falcon-ggml/
# github.com/go-skynet/LocalAI/cmd/grpc/falcon-ggml
/usr/local/go/pkg/tool/linux_amd64/link: running g++ failed: exit status 1
/usr/bin/ld: cannot find -lOpenCL
/usr/bin/ld: cannot find -lclblast
collect2: error: ld returned 1 exit status

make: *** [Makefile:387: backend-assets/grpc/falcon-ggml] Error 1
```

env:

```
BUILD_TYPE=clblas
CLBLAST_DIR=/mnt/clblast
```

docker-compose:

```yaml
  localai:
    image: quay.io/go-skynet/local-ai:master
...
    build:
      context: ${MOUNT_DOCKER_DATA}/LocalAI/git/
      dockerfile: Dockerfile
    ports:
      - 8888:8080
    env_file:
      - localai/env
    volumes:
      - ${MOUNT_DOCKER_DATA}/LocalAI/models:/models
      - ${MOUNT_DOCKER_DATA}/LocalAI/clblast:/mnt/clblast
...
```

clblast:

```
/opt/docker-data/LocalAI/clblast

ls -1

bin/
include/
lib/

ls -1 bin/

clblast_sample_cache_c*
clblast_sample_dgemv_c*
clblast_sample_dtrsm*
clblast_sample_haxpy_c*
clblast_sample_samax_c*
clblast_sample_sasum_c*
clblast_sample_sgemm*
clblast_sample_sgemm_batched*
clblast_sample_sgemm_c*
clblast_sample_tuning_api*
clblast_tuner_copy_fast*
clblast_tuner_copy_pad*
clblast_tuner_invert*
clblast_tuner_routine_xgemm*
clblast_tuner_routine_xtrsv*
clblast_tuner_transpose_fast*
clblast_tuner_transpose_pad*
clblast_tuner_xaxpy*
clblast_tuner_xconvgemm*
clblast_tuner_xdot*
clblast_tuner_xgemm*
clblast_tuner_xgemm_direct*
clblast_tuner_xgemv*
clblast_tuner_xger*

ls -1 lib/

cmake/
libclblast.so -> libclblast.so.1
libclblast.so.1 -> libclblast.so.1.6.1
libclblast.so.1.6.1
pkgconfig/
```

I have run into the exact same issue, but using make manually on a Ubuntu 22.04 system (no docker). I used the following steps (following https://localai.io/basics/build/index.html where applicable):

```bash
mkdir -p ~/src && cd ~/src

INSTALL_PREFIX=~/src/install
mkdir -p ${INSTALL_PREFIX}

git clone --recurse-submodules https://github.com/KhronosGroup/OpenCL-SDK.git

mkdir -p OpenCL-SDK/build
cd OpenCL-SDK/build

cmake .. -DBUILD_DOCS=OFF -DBUILD_EXAMPLES=OFF -DBUILD_TESTING=OFF -DOPENCL_SDK_BUILD_SAMPLES=OFF -DOPENCL_SDK_TEST_SAMPLES=OFF
cmake --build . --config Release -j 7
cmake --install . --prefix ${INSTALL_PREFIX}

cd ~/src

git clone https://github.com/CNugteren/CLBlast.git

mkdir -p CLBlast/build
cd CLBlast/build

cmake .. -DBUILD_SHARED_LIBS=OFF -DTUNERS=OFF -DOPENCL_ROOT=${INSTALL_PREFIX}
cmake --build . --config Release -j 7
cmake --install . --prefix ${INSTALL_PREFIX}

cd ~/src

git clone --recurse-submodules -j8 https://github.com/mudler/LocalAI.git
cd LocalAI

CMAKE_ARGS="-DLLAMA_F16C=OFF -DLLAMA_AVX512=OFF -DLLAMA_AVX2=OFF -DLLAMA_FMA=OFF" CLBLAST_DIR=${INSTALL_PREFIX}/lib/cmake/CLBlast/ make BUILD_TYPE=clblas GO_TAGS=tts build
```

The full log is [here](https://pastebin.com/3QPBTSiD), but fails at the same point as @sammcj 

I also tried building with the following (making sure to make clean first):
```
CMAKE_ARGS="-DLLAMA_F16C=OFF -DLLAMA_AVX512=OFF -DLLAMA_AVX2=OFF -DLLAMA_FMA=OFF -DLLAMA_CLBLAST=ON -DCLBLAST_DIR=${INSTALL_PREFIX}/lib/cmake/CLBlast/" CLBLAST_DIR=${INSTALL_PREFIX}/lib make BUILD_TYPE=clblas GO_TAGS=tts build
```
And also tried  with `CLBLAST_DIR=${INSTALL_PREFIX}/lib/cmake/CLBlast/`

```
~/src/install/lib$ ls -1
cmake/
libclblast.a
libOpenCLExt.a
libOpenCL.so
libOpenCL.so.1
libOpenCL.so.1.2
libOpenCLUtils.a
libOpenCLUtilsCpp.a
pkgconfig/
```

I also tried removing `falcon-ggml` from the makefile to see if I could work around it, but still ran into linker issues with OpenCL and clblast with the next target in the makefile:

```
...
CGO_LDFLAGS=" -lOpenCL -lclblast" C_INCLUDE_PATH=~/src/LocalAI/sources/go-bert LIBRARY_PATH=~/src/LocalAI/sources/go-bert \
go build -ldflags " -X "github.com/go-skynet/LocalAI/internal.Version=v2.0.0-16-g89ff123" -X "github.com/go-skynet/LocalAI/internal.Commit=89ff12309daac0a9a4f6e85b7cfc3833995d4e82"" -tags "tts" -o backend-assets/grpc/bert-embeddings ./backend/go/llm/bert/
# github.com/go-skynet/go-bert.cpp
# github.com/go-skynet/LocalAI/backend/go/llm/bert
/usr/local/go/pkg/tool/linux_amd64/link: running g++ failed: exit status 1
/usr/bin/ld: cannot find -lOpenCL
/usr/bin/ld: cannot find -lclblast
/usr/bin/ld: cannot find -lOpenCL
/usr/bin/ld: cannot find -lclblast
/usr/bin/ld: cannot find -lOpenCL
/usr/bin/ld: cannot find -lclblast
/usr/bin/ld: cannot find -lOpenCL
/usr/bin/ld: cannot find -lclblast
collect2: error: ld returned 1 exit status

make: *** [Makefile:512: backend-assets/grpc/bert-embeddings] Error 1
```

--- a/docs/content/build/_index.en.md
+++ b/docs/content/build/_index.en.md
@@ -147,20 +147,24 @@ make BUILD_TYPE=cublas build
 
 More informations available in the upstream PR: https://github.com/ggerganov/llama.cpp/pull/1412
 
-#### Hipblas (AMD GPU)
 
-AMD GPU Acceleration
+#### Hipblas (AMD GPU with ROCm on Arch Linux)
 
-Requirement: ROCm
-
+Packages:
 ```
-make BUILD_TYPE=hipblas build
+pacman -S base-devel git rocm-hip-sdk rocm-opencl-sdk opencv clblast grpc
 ```
 
-Specific GPU targets can be specified with `GPU_TARGETS`:
-  
+Library links:
 ```
-make BUILD_TYPE=hipblas GPU_TARGETS=gfx90a build
+export CGO_CFLAGS="-I/usr/include/opencv4"
+export CGO_CXXFLAGS="-I/usr/include/opencv4"
+export CGO_LDFLAGS="-L/opt/rocm/hip/lib -lamdhip64 -L/opt/rocm/lib -lOpenCL -L/usr/lib -lclblast -lrocblas -lhipblas -lrocrand -lomp -O3 --rtlib=compiler-rt -unwindlib=libgcc -lhipblas -lrocblas --hip-link"
+```
+
+Build:
+```
+make BUILD_TYPE=hipblas GPU_TARGETS=gfx1030
 ```
 
 #### ClBLAS

